File: dwarf/targets/cardano-node/src/DecodeBlockFetch.hs
{-# LANGUAGE OverloadedStrings #-}
-- | cardano-node-mini-protocol-decode-blockfetch shim.
-- Decodes one BlockFetch mini-protocol wire-format message.
module Main (main) where

import qualified Codec.CBOR.Decoding              as CBOR
import qualified Codec.CBOR.Read                  as CBOR
import qualified Data.ByteString                  as BS
import qualified Data.ByteString.Lazy             as BSL
import           System.Exit                      (ExitCode (..), exitSuccess, exitWith)

decodePoint :: CBOR.Decoder s ()
decodePoint = do
  len <- CBOR.decodeListLen
  case len of
    0 -> pure ()
    2 -> do
      _slot <- CBOR.decodeWord64
      hash <- CBOR.decodeBytes
      if BS.length hash == 32
         then pure ()
         else fail ("point hash length " ++ show (BS.length hash) ++ ", expected 32")
    _ -> fail ("point length " ++ show len ++ ", expected 0 or 2")

decoder :: CBOR.Decoder s ()
decoder = do
  len <- CBOR.decodeListLen
  key <- CBOR.decodeWord
  case (len, key) of
    (3, 0) -> decodePoint >> decodePoint
    (1, 1) -> pure ()
    (1, 2) -> pure ()
    (1, 3) -> pure ()
    (2, 4) -> do
      tag <- CBOR.decodeTag
      if tag == 24
         then pure ()
         else fail ("block tag " ++ show tag ++ ", expected 24")
      _bytes <- CBOR.decodeBytes
      pure ()
    (1, 5) -> pure ()
    _      -> fail ("unexpected blockfetch message (len, key) = (" ++ show len ++ ", " ++ show key ++ ")")

main :: IO ()
main = do
  bytes <- BSL.getContents
  case CBOR.deserialiseFromBytes decoder bytes of
    Left err -> do
      putStrLn ("ERR " ++ show err)
      exitWith (ExitFailure 1)
    Right (rest, ()) | BSL.null rest -> do
      putStrLn "OK"
      exitSuccess
    Right _ -> do
      putStrLn "ERR trailing bytes after message"
      exitWith (ExitFailure 1)

File: dwarf/targets/amaru/src/bin/decode_blockfetch.rs
// amaru-mini-protocol-decode-blockfetch shim.
// Decodes one BlockFetch mini-protocol wire-format message:
//   RequestRange = [0, from_point, through_point]
//   ClientDone   = [1]
//   StartBatch   = [2]
//   NoBlocks     = [3]
//   Block        = [4, tag(24) block_bytes]
//   BatchDone    = [5]

use std::io::{self, Read};
use std::process::ExitCode;

fn decode_point(d: &mut minicbor::Decoder<'_>) -> Result<(), String> {
    let len = d
        .array()
        .map_err(|e| format!("point array: {e}"))?
        .ok_or_else(|| "indefinite-length point not allowed".to_string())?;
    match len {
        0 => Ok(()),
        2 => {
            let _slot = d.u64().map_err(|e| format!("point slot: {e}"))?;
            let hash = d.bytes().map_err(|e| format!("point hash: {e}"))?;
            if hash.len() != 32 {
                return Err(format!("point hash length {}, expected 32", hash.len()));
            }
            Ok(())
        }
        _ => Err(format!("point length {len}, expected 0 or 2")),
    }
}

fn decode(data: &[u8]) -> Result<(), String> {
    let mut d = minicbor::Decoder::new(data);
    let len = d
        .array()
        .map_err(|e| format!("message array: {e}"))?
        .ok_or_else(|| "indefinite-length blockfetch message not allowed".to_string())?;
    let key: u32 = d.u32().map_err(|e| format!("message key: {e}"))?;
    match (len, key) {
        (3, 0) => {
            decode_point(&mut d)?;
            decode_point(&mut d)?;
        }
        (1, 1) | (1, 2) | (1, 3) | (1, 5) => {}
        (2, 4) => {
            let tag = d.tag().map_err(|e| format!("block tag: {e}"))?;
            if u64::from(tag) != 24 {
                return Err(format!("block tag {}, expected 24", u64::from(tag)));
            }
            let _bytes = d.bytes().map_err(|e| format!("block bytes: {e}"))?;
        }
        _ => {
            return Err(format!(
                "unexpected blockfetch message (len, key) = ({len}, {key})"
            ))
        }
    }
    if d.position() != data.len() {
        return Err(format!("trailing bytes at position {}", d.position()));
    }
    Ok(())
}

fn main() -> ExitCode {
    let mut buf = Vec::new();
    if let Err(e) = io::stdin().read_to_end(&mut buf) {
        println!("ERR stdin read: {e}");
        return ExitCode::from(1);
    }
    match decode(&buf) {
        Ok(()) => {
            println!("OK");
            ExitCode::from(0)
        }
        Err(msg) => {
            println!("ERR {msg}");
            ExitCode::from(1)
        }
    }
}
